### PR TITLE
MRG: Fix Scaler

### LINKED ISSUE #3665: ENH: decoding robust option in Scaler
Currently the `mne.decoding.Scaler` takes the mean and std across channels to normalize the data.

We should have an option to either accept and sklearn.transformer, a dict a string, so that we can get 
- default values for meeg, eeg etc e.g: `Scaler(info, method='default')` would apply `mne.cov._check_scalings_user()`
- data-estimated standard scaling: i.e. current behavior i.e. `X/std-mean` over time
- data-estimated robust scaling : e.g. `Scaler(method=sklearn.preprocessing.RobustScaler())`

@kaichogami that should be an easy one if you're interested

### PR DESCRIPTION
**Pull request by Eric89GXL**
Closes #3665.

Ready for review/merge from my end. @kingjr can you verify this is the functionality you were thinking of? The tests should make it pretty clear.

### CODE REVIEW

File: mne/decoding/tests/test_transformer.py [hunk 1]
```diff
@@ -39,26 +39,46 @@ def test_scaler():
...
+        if method is None or isinstance(method, dict):
+            sd = DEFAULTS['scalings'] if method is None else method
+            stds = np.zeros(len(picks))
+            for key in ('mag', 'grad'):
+                stds[pick_types(epochs.info, meg=key)] = 1. / sd[key]
+            stds[pick_types(epochs.info, meg=False, eeg=True)] = 1. / sd['eeg']
+            means = np.zeros(len(epochs.ch_names))
+        elif method == 'mean':
+            stds = np.array([np.std(e) for e in epochs_data_t])
+            means = np.array([np.mean(e) for e in epochs_data_t])
```
Comment (kingjr) [inline mne/decoding/tests/test_transformer.py]: would call `e` => `channel` or `ch`

But yes, it's the intended behavior IMO.

File: mne/decoding/tests/test_transformer.py [hunk 2]
```diff
@@ -39,26 +39,46 @@ def test_scaler():
...
+            continue
+        if method == 'mean' and not check_version('sklearn', ''):
+            assert_raises(ImportError, Scaler, info, method)
+            continue
+        scaler = Scaler(info, method)
+        X = scaler.fit_transform(epochs_data, y)
+        assert_equal(X.shape, epochs_data.shape)
+        if method is None or isinstance(method, dict):
+            sd = DEFAULTS['scalings'] if method is None else method
+            stds = np.zeros(len(picks))
```
Comment (kingjr) [inline mne/decoding/tests/test_transformer.py]: np.zeros_like

File: mne/decoding/tests/test_transformer.py [hunk 3]
```diff
@@ -39,26 +39,46 @@ def test_scaler():
...
+                stds[pick_types(epochs.info, meg=key)] = 1. / sd[key]
+            stds[pick_types(epochs.info, meg=False, eeg=True)] = 1. / sd['eeg']
+            means = np.zeros(len(epochs.ch_names))
+        elif method == 'mean':
+            stds = np.array([np.std(e) for e in epochs_data_t])
+            means = np.array([np.mean(e) for e in epochs_data_t])
+        else:  # median
+            percs = np.array([np.percentile(e, [25, 50, 75])
+                              for e in epochs_data_t])
+            stds = percs[:, 2] - percs[:, 0]
```
Comment (kingjr) [inline mne/decoding/tests/test_transformer.py]: is that right? Aren't we doing an absolute median deviation?

Comment (Eric89GXL) [inline mne/decoding/tests/test_transformer.py]: That's the interquartile range (75th - 25th) which is what RobustScaler does in sklearn

File: mne/decoding/tests/test_transformer.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/decoding/tests/test_transformer.py]: `picks` will be `int` so `np.zeros_like(picks)` is incorrect here

**Comment by Eric89GXL** [comment]
@kingjr you happy, then?

File: mne/decoding/tests/test_transformer.py (hunk 2 above)
Comment (kingjr) [inline mne/decoding/tests/test_transformer.py]: I don't understand, `int` has no len(). But that's probably not worth a discussion..

**Comment by kingjr** [comment]
yes!

**Comment by agramfort** [comment]
@jasmainak I'll let you take a look first

File: mne/decoding/tests/test_transformer.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/decoding/tests/test_transformer.py]: you mean `len(picks)`? `picks` is a list of int here

Comment (Eric89GXL) [inline mne/decoding/tests/test_transformer.py]: ... meant to say that `picks` will be `dtype=int`, sorry

File: mne/decoding/transformer.py [hunk 4]
```diff
@@ -13,42 +13,113 @@
...
+    scalings : dict, string, defaults to None.
+        Scaling method to be applied to data channel wise.
+
+        * if scalings is None (default), scales mag by 1e15, grad by 1e13,
+          and eeg by 1e6.
+        * if scalings is :class:`dict`, keys are channel types and values
+          are scale factors.
+        * if ``scalings=='median'``,
+          :class:`sklearn.preprocessing.RobustScaler`
+          is used (requires sklearn version 0.17+).
```
Comment (jasmainak) [inline mne/decoding/transformer.py]: I don't quite follow the logic of having a wrapper around `sklearn`. Why is this done? Couldn't the user directly use `sklearn`

File: mne/decoding/transformer.py [hunk 5]
```diff
@@ -13,42 +13,113 @@
...
+        * if ``scalings=='mean'``,
+          :class:`sklearn.preprocessing.StandardScaler`
```
Comment (jasmainak) [inline mne/decoding/transformer.py]: same question here ...

File: mne/decoding/transformer.py [hunk 6]
```diff
@@ -13,42 +13,113 @@
...
+
+    def inverse_transform(self, X, y=None):
+        return X * self.std_
+
+    def fit_transform(self, X, y=None):
+        return self.fit(X, y).transform(X)
+
+
+def _sklearn_reshape_apply(func, return_result, X, *args):
+    """Reshape epochs and apply function."""
```
Comment (jasmainak) [inline mne/decoding/transformer.py]: Why not use `Vectorizer` in a pipeline?

File: mne/decoding/transformer.py [hunk 7]
```diff
@@ -13,42 +13,113 @@
...
+from ..cov import _check_scalings_user
 
 
-class Scaler(TransformerMixin):
-    u"""Standardize data across channels.
+class _ConstantScaler():
+    """Scale channel types using constant values."""
+
+    def __init__(self, info, scalings, do_scaling=True):
+        self._scalings = _check_scalings_user(scalings)
```
Comment (jasmainak) [inline mne/decoding/transformer.py]: I think the `__init__` should not have any logic for it to work with `GridSearchCV` ...

File: mne/decoding/tests/test_transformer.py [hunk 8]
```diff
@@ -39,26 +39,46 @@ def test_scaler():
...
-    # Test inverse_transform
-    Xi = scaler.inverse_transform(X, y)
-    assert_array_almost_equal(epochs_data, Xi)
-
-    for kwargs in [{'with_mean': False}, {'with_std': False}]:
-        scaler = Scaler(epochs.info, **kwargs)
-        scaler.fit(epochs_data, y)
-        assert_array_almost_equal(
-            X, scaler.inverse_transform(scaler.transform(X)))
+    methods = (None, dict(mag=5, grad=10, eeg=20), 'mean', 'median')
```
Comment (jasmainak) [inline mne/decoding/tests/test_transformer.py]: I think you should throw an error if there are channels other than `mag`, `grad` or `eeg`. It would be weird to scale only a subset of channels ...

File: mne/decoding/transformer.py [hunk 9]
```diff
@@ -13,42 +13,113 @@
...
+    X = func(X, *args)
+    if return_result:
+        X.shape = (orig_shape[0], orig_shape[2], orig_shape[1])
+        X = X.transpose(0, 2, 1)
+        return X
+
 
-    By default, this makes each time point (within each epoch) have
-    μ=0, σ=1.
+class Scaler(TransformerMixin, BaseEstimator):
```
Comment (jasmainak) [inline mne/decoding/transformer.py]: why both `TransformerMixin` and `BaseEstimator` ?

**Comment by jasmainak** [comment]
okay, I'm done with first pass of my review

File: doc/manual/decoding.rst [hunk 10]
```diff
@@ -17,9 +17,11 @@ Basic Estimators
...
+This will scale data across channels. Each channel type (mag, grad or eeg) is treated separately. Users can specify the method to scale with. During training time, ``scikit-learn`` objects are created if scaling method is ``StandardScaler`` or ``RobustScaler``. The ``transform`` method is called to transform the training set. To perform both the ``fit`` and ``transform`` operations in a single call, the ``fit_transform`` method may be used. During test time, the stored scaling values are used in the ``transform`` method. To recover the original data, you can use ``inverse_transform``.
 
-.. note:: This is different from the ``StandarScaler`` estimator offered by Scikit-Learn. The ``StandardScaler`` standardizes each feature, whereas the ``Scaler`` object standardizes by channel type.
+.. note:: For working of every scaling method, atleast scikit-learn version 0.18 is required.
+
+.. note:: This is different from directly applying ``StandarScaler`` or ``RobustScaler``  estimator offered by Scikit-Learn. The ``StandardScaler`` or ``RobustScaler`` scales each feature, whereas the ``Scaler`` object scales by channel type.
```
Comment (jasmainak) [inline doc/manual/decoding.rst]: Ah, I saw this now.

File: doc/manual/decoding.rst [hunk 11]
```diff
@@ -17,9 +17,11 @@ Basic Estimators
...
+This will scale data across channels. Each channel type (mag, grad or eeg) is treated separately. Users can specify the method to scale with. During training time, ``scikit-learn`` objects are created if scaling method is ``StandardScaler`` or ``RobustScaler``. The ``transform`` method is called to transform the training set. To perform both the ``fit`` and ``transform`` operations in a single call, the ``fit_transform`` method may be used. During test time, the stored scaling values are used in the ``transform`` method. To recover the original data, you can use ``inverse_transform``.
 
-.. note:: This is different from the ``StandarScaler`` estimator offered by Scikit-Learn. The ``StandardScaler`` standardizes each feature, whereas the ``Scaler`` object standardizes by channel type.
+.. note:: For working of every scaling method, atleast scikit-learn version 0.18 is required.
```
Comment (jasmainak) [inline doc/manual/decoding.rst]: I think I saw 0.17 in the error messages ...

File: mne/decoding/transformer.py [hunk 12]
```diff
@@ -13,42 +13,113 @@
...
+        if scalings is None or isinstance(scalings, dict):
+            self._scaler = _ConstantScaler(info, scalings, self.with_std)
+        elif scalings == 'mean':
+            from sklearn.preprocessing import StandardScaler
+            self._scaler = StandardScaler(self.with_mean, self.with_std)
+        else:  # scalings == 'median':
+            if not check_version('sklearn', '0.17'):
+                raise ValueError("median requires version 0.17 of "
+                                 "sklearn library")
+            from sklearn.preprocessing import RobustScaler
```
Comment (jasmainak) [inline mne/decoding/transformer.py]: since the API of `RobustScaler` and `Scaler` is the same, you could also ask the user to simply pass the estimator object, so you avoid the version checks etc.

File: mne/decoding/transformer.py [hunk 13]
```diff
@@ -13,42 +13,113 @@
...
+    info : instance of Info | None
+        The measurement info. Only necessary if ``scalings`` is a dict or
+        None.
+    scalings : dict, string, defaults to None.
+        Scaling method to be applied to data channel wise.
+
+        * if scalings is None (default), scales mag by 1e15, grad by 1e13,
+          and eeg by 1e6.
+        * if scalings is :class:`dict`, keys are channel types and values
+          are scale factors.
```
Comment (jasmainak) [inline mne/decoding/transformer.py]: is there a good reason to allow scales other than the defaults? 

also, since we the scales are constant, it's not really an estimator object. It's not estimating anything from the data ... you don't even need a `fit` and `transform`.

File: mne/decoding/transformer.py (hunk 9 above)
Comment (Eric89GXL) [inline mne/decoding/transformer.py]: I don't know, it was like this when I inherited the code

File: mne/decoding/transformer.py (hunk 7 above)
Comment (Eric89GXL) [inline mne/decoding/transformer.py]: I'm not sure what you mean. Inputs shouldn't be vetted or processed?

File: mne/decoding/transformer.py (hunk 6 above)
Comment (Eric89GXL) [inline mne/decoding/transformer.py]: Don't know, @kingjr?

File: doc/manual/decoding.rst (hunk 10 above)
Comment (Eric89GXL) [inline doc/manual/decoding.rst]: Actually this isn't true. `Scaler` only really operates on all channel types the same way for the `dict | None` options, for the others it's basically a wrapper.

File: mne/decoding/transformer.py (hunk 12 above)
Comment (Eric89GXL) [inline mne/decoding/transformer.py]: @agramfort requested string parsing like this (your suggestion was @kingjr's original proposal)

File: mne/decoding/transformer.py (hunk 13 above)
Comment (Eric89GXL) [inline mne/decoding/transformer.py]: I assume to be `sklearn` compatible, we need these methods. And `fit` needs to set the `name_` attributes, and `transform` needs to actually modify the data that is passed.

File: mne/decoding/transformer.py (hunk 6 above)
Comment (Eric89GXL) [inline mne/decoding/transformer.py]: ...oh, I think I can explain the difference. `Vectorizer` would treat each channel *and time point* independently since they are all features, just scaling each one across its set of epochs (right?). This will scale each channel, including all time points from all epochs from it together.

File: mne/decoding/transformer.py (hunk 7 above)
Comment (jasmainak) [inline mne/decoding/transformer.py]: They should be, but you should move the logic to `fit` I think. From the `scikit-learn` documentation:

> There should be no logic, not even input validation, and the parameters should not be changed. The corresponding logic should be put where the parameters are used, typically in fit

File: mne/decoding/transformer.py (hunk 12 above)
Comment (jasmainak) [inline mne/decoding/transformer.py]: hmm ... okay. I'll ask @agramfort IRL tomorrow if I see him

File: mne/decoding/tests/test_transformer.py (hunk 8 above)
Comment (Eric89GXL) [inline mne/decoding/tests/test_transformer.py]: That is taken care of in the `StandardScaler.fit` method, but I guess I could add a test

File: mne/decoding/transformer.py (hunk 13 above)
Comment (jasmainak) [inline mne/decoding/transformer.py]: yeah, but you need to be `sklearn` compatible only if you use it in a pipeline or grid search. Here, you just scale X beforehand by constant (without caring about training or test set). That way, you don't even need a class ...

Comment (Eric89GXL) [inline mne/decoding/transformer.py]: > yeah, but you need to be sklearn compatible only if you use it in a pipeline or grid search. 

I figured the point was to be able to put it into a `Pipeline` easily

File: doc/manual/decoding.rst (hunk 10 above)
Comment (jasmainak) [inline doc/manual/decoding.rst]: okay, that's a bit confusing ...

I was expecting `StandardScaler` and `RobustScaler` to somehow scale by channel type. If they're not doing that, we don't even need them.

Comment (Eric89GXL) [inline doc/manual/decoding.rst]: I was wrong, they are still different (see other reply). I'm also about to push an updated `manual` entry that hopefully clarifies it.

File: mne/decoding/transformer.py (hunk 6 above)
Comment (jasmainak) [inline mne/decoding/transformer.py]: yes, `Vectorizer` basically concatenates the channels into a single vector. So, your epochs is a 2D array. I didn't get this:

> This will scale each channel, including all time points from all epochs from it together.

File: mne/decoding/transformer.py (hunk 13 above)
Comment (jasmainak) [inline mne/decoding/transformer.py]: yeah, but you put things in `Pipeline` only if you need to estimate something on train set and then transform or predict on test set. If it's constant across all trials, you can just do it without `Pipeline`. 

I don't know if you want to keep it for the sake of "completeness" ... but the `RobustScaler` and `Scaler` also feel like just wrappers

File: mne/decoding/transformer.py (hunk 9 above)
Comment (jasmainak) [inline mne/decoding/transformer.py]: okay, no objection. Just curious

File: mne/decoding/transformer.py (hunk 6 above)
Comment (kingjr) [inline mne/decoding/transformer.py]: > This will scale each channel, including all time points from all epochs from it together.

yes

@jasmainak Perhaps this formulation isn't entirely clear indeed. I would go for pseudo code or equation to make it clearer.

File: doc/manual/decoding.rst (hunk 10 above)
Comment (jasmainak) [inline doc/manual/decoding.rst]: I'm lost which reply you are talking about. So, do all of them scale by channel type?

File: mne/decoding/transformer.py (hunk 12 above)
Comment (kingjr) [inline mne/decoding/transformer.py]: I think alex pointed that it was a YAGNI indeed. I think I tend to agree.

File: mne/decoding/transformer.py (hunk 6 above)
Comment (Eric89GXL) [inline mne/decoding/transformer.py]: I'm just looking into this stuff now, but it looks like `Vectorizer` doesn't do what you say. It says it collapses across all features (e.g., channels x time) so you end up with `(n_epochs, n_features)`. So if you do `StandardScaler` on top of that, you standardize each *feature* (e.g., a single time point for a single channel) across its epochs, whereas this standardizes each *channel* across all epochs and time points.

I'll push the doc updates, hopefully that clarifies things

**Comment by jasmainak** [comment]
@Eric89GXL is there an example that could be perhaps updated? I think it might be possible to demo `RobustScaler` on sample because of the bad channel

File: mne/decoding/transformer.py (hunk 6 above)
Comment (jasmainak) [inline mne/decoding/transformer.py]: > I'm just looking into this stuff now, but it looks like Vectorizer doesn't do what you say. It says it collapses across all features (e.g., channels x time) so you end up with (n_epochs, n_features). So if you do StandardScaler on top of that, you standardize each feature (e.g., a single time point for a single channel) across its epochs,

Indeed that is what I meant. Maybe I formulated it differently ...

**Comment by Eric89GXL** [comment]
> is there an example that could be perhaps updated?

Perhaps in another PR, this is really a bug fix

**Comment by Eric89GXL** [comment]
Okay, docs have been clarified

File: doc/manual/decoding.rst [hunk 14]
```diff
@@ -17,9 +18,11 @@ Basic Estimators
...
+
+.. note:: This is different from directly applying :class:`sklearn.preprocessing.StandardScaler` or :class:`sklearn.preprocessing.RobustScaler` offered by scikit-learn_. The ``StandardScaler`` and ``RobustScaler`` scale each *classification feature*, e.g. each time point for each channel, with mean and standard deviation computed across epochs, whereas ``Scaler`` scales each *channel* using mean and standard deviation computed across all of its time points and epochs.
```
Comment (jasmainak) [inline doc/manual/decoding.rst]: okay, now I get what's going on. But still, I find this weird.. If you scale each channel differently, aren't you going to kill the spatial information? The scale should be same for all channels in one channel type, no?

Comment (Eric89GXL) [inline doc/manual/decoding.rst]: Yeah it will distort the spatial profile. @kingjr?

**Comment by agramfort** [comment]
let's say it's a bug fix. We'll iterate during the sprint.
